File: java/nitf/src/main/java/nitf/TRE.java
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 * 
 * (C) Copyright 2004 - 2008, General Dynamics - Advanced Information Systems
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public 
 * License along with this program; if not, If not, 
 * see <http://www.gnu.org/licenses/>.
 *
 */

package nitf;

import java.io.PrintStream;

/**
 * A representation of the NITF Tagged Record Extensions (TRE)
 */
public final class TRE extends DestructibleObject
{

    /**
     * Specifies that the length of the TRE should be computed by the library, and not specified by the user at
     * construction time
     */
    public static final int DEFAULT_LENGTH = 0;

    /**
     * Creates a new TRE of the given type
     * 
     * @param tag
     *            the type of TRE to create
     * @throws NITFException
     *             if tag is an unkown type
     */
    public TRE(String tag) throws NITFException
    {
        this(tag, DEFAULT_LENGTH);
    }

    /**
     * Creates a new TRE of the given type, with the given length
     * 
     * @param tag
     *            the type of TRE to create
     * @param size
     *            the size of the TRE, or DEFAULT_LENGTH
     * @throws NITFException
     *             if tag is an unkown type
     */
    public TRE(String tag, int size) throws NITFException
    {
        this(tag, size, null);
    }

    /**
     * Creates a new TRE of the given type, and uses the given TREDescription[] for the description of the TRE. This
     * should only be used if you have created a valid description. Passing in null for the description is the same as
     * calling the TRE(String tag) constructor.
     * 
     * @param tag
     *            the type of TRE to create
     * @param description
     *            TREDescription
     * @throws NITFException
     */
    public TRE(String tag, int size, TREDescription[] description)
            throws NITFException
    {
        if (!PluginRegistry.canHandleTRE(tag))
            throw new NITFException(
                    "TRE Handler cannot be found for this TRE: " + tag);

        if (!construct(tag, size, description))
        {
            throw new NITFException("Could not create TRE with tag: [" + tag
                    + "]");
        }
    }

    /**
     * @see NITFObject#NITFObject(long)
     */
    TRE(long address)
    {
        super(address);
    }

    protected native synchronized void destructMemory();

    private native boolean construct(String tag, int size,
            TREDescription[] description) throws NITFException;

    /**
     * Returns the length of this TRE
     * 
     * @return
     */
    public native synchronized int getLength();

    /**
     * Sets the length of the TRE
     * 
     * @param length
     * @throws NITFException
     */
    public native void setLength(int length) throws NITFException;

    /**
     * Computes the current length of the TRE by iterating over it
     * 
     * @return
     */
    public native synchronized int computeLength();

    /**
     * Returns the TREDescription array for this TRE
     * 
     * @return
     */
    public native synchronized TREDescription[] getTREDescription();

    /**
     * Returns the TRE identifier tag
     * 
     * @return the TRE identifier tag
     */
    public native synchronized String getTag();

    /**
     * Returns true if the given field tag exists within the TRE The tag is the identifier for the field. Check out the
     * TRE Descriptions for a listing of the field tags for each TRE. Note - Fields that loop will have a tag TAG[i],
     * such that i is the number in the loop, starting with 0
     * 
     * @param tag
     *            the identifier for the field
     * @return true if the field exists, false otherwise
     */
    public native synchronized boolean exists(String tag);

    /**
     * Returns the Field associated with the given tag
     * 
     * @param tag
     *            the identifier for the field
     * @throws NITFException
     *             if an error occurs or no field with with the given tag exists
     */
    public native synchronized Field getField(String tag) throws NITFException,
            NoSuchFieldException;

    /**
     * Attempts to set the value of the field referenced by tag to the data given. Throws an NITFException if an error
     * occurs.
     * 
     * @param tag
     *            the identifier for the field
     * @param data
     *            the data to use
     * @throws NITFException
     *             if an error occurs
     */
    public native synchronized boolean setField(String tag, byte[] data)
            throws NITFException;

    /**
     * Returns true if the TRE is sane, i.e. it contains all of the values that should be there, based on the
     * description of the TRE. If it is not sane, it returns false
     * 
     * @return true if the TRE is sane, false otherwise
     */
    public native synchronized boolean isSane();

    /**
     * Prints the contents of the TRE to the given PrintStream
     * 
     * @param stream
     *            PrintStream to print to
     * @throws NITFException
     */
    public synchronized void print(PrintStream stream) throws NITFException
    {
        stream.println("\n---------------" + getTag() + "---------------");
        final TREField[] fields = getFields();
        for (int i = 0; i < fields.length; i++)
        {
            TREField field = fields[i];

            stream
                    .println(field.getDescription().getTag()
                            + " ("
                            + field.getDescription().getLabel()
                            + ") = ["
                            + (field.getField().getType() == FieldType.NITF_BINARY ? ("<binary stuff, length = "
                                    + field.getField().getLength() + ">")
                                    : field.getField().toString()) + "]");
        }
        stream.println("------------------------------------");
    }

    /**
     * Attempts to re-set the description to the one provided by the plugin for the given tag This function also sets
     * the tag for the tre to the parameter
     * 
     * @param tag
     *            the main TRE identifier
     * @param description
     *            OPTIONAL description to set it to
     * @return true if success, false otherwise
     */
    // private native synchronized boolean setDescription(String tag,
    // TREDescription[] description);
    /**
     * Return all of the Fields that are a part of the TRE, in order
     * 
     * @return Field array
     * @throws NITFException
     */
    public native synchronized TREField[] getFields() throws NITFException;

    /**
     * Contains an association between a TRE Field and Description
     */
    public static class TREField
    {
        protected TREDescription description = null;

        protected Field field = null;

        protected String qualifiedTag = null;

        protected TREField(Field field, TREDescription description)
        {
            this.description = description;
            this.field = field;
        }

        /**
         * Returns the TREDescription
         * 
         * @return TREDescription
         */
        public TREDescription getDescription()
        {
            return description;
        }

        /**
         * Returns the Field
         * 
         * @return Field
         */
        public Field getField()
        {
            return field;
        }

        /**
         * @return Returns the qualifiedTag.
         */
        public String getQualifiedTag()
        {
            return qualifiedTag;
        }

        /**
         * @param qualifiedTag
         *            The qualifiedTag to set.
         */
        protected void setQualifiedTag(String qualifiedTag)
        {
            this.qualifiedTag = qualifiedTag;
        }

    }

}

File: externals/nitro/modules/c++/nitf/source/TRE.cpp
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 *
 * (C) Copyright 2004 - 2014, MDA Information Systems LLC
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, If not,
 * see <http://www.gnu.org/licenses/>.
 *
 */

#include "nitf/TRE.hpp"
#include <string.h>
#include "nitf/TREUtils.h"

#include "gsl/gsl.h"

using namespace nitf;

TRE::TRE(const TRE& x)
{
    *this = x;
}

TRE& TRE::operator=(const TRE& x)
{
    if (&x != this)
        setNative(x.getNative());
    return *this;
}

TRE::TRE(nitf_TRE* x)
{
    setNative(x);
    getNativeOrThrow();
}

TRE::TRE(NITF_DATA* x) : TRE(static_cast<nitf_TRE*>(x))
{
}

TRE& TRE::operator=(NITF_DATA* x)
{
    setNative(static_cast<nitf_TRE*>(x));
    getNativeOrThrow();
    return *this;
}

TRE::TRE(const char* tag, const char* id) : TRE(tag, id, nullptr)
{
}
TRE::TRE(const char* tag) : TRE(tag, "")
{
}
TRE::TRE(const std::string& tag) : TRE(tag, "")
{
}
TRE::TRE(const std::string& tag, const std::string& id, std::nullptr_t) : TRE(create(tag, id, error))
{
    setManaged(false);
}

nitf::TRE TRE::clone() const
{
    nitf::TRE dolly(nitf_TRE_clone(getNativeOrThrow(), &error));
    dolly.setManaged(false);
    return dolly;
}

TRE::Iterator TRE::begin() const
{
    nitf_TREEnumerator* iter = nitf_TRE_begin(getNativeOrThrow(), &error);
    if (!iter)
        throw nitf::NITFException(Ctxt("Invalid TRE: " + getTag()));
    return TRE::Iterator(iter);
}

TRE::Iterator TRE::end() const noexcept
{
    return TRE::Iterator();
}

nitf::Field TRE::getField(const std::string& key) const
{
    nitf_Field* field = nitf_TRE_getField(getNativeOrThrow(), key.c_str());
    if (!field)
        throw except::NoSuchKeyException(
                Ctxt(FmtX("Field does not exist in TRE: %s", key.c_str())));
    return nitf::Field(field);
}

void TRE::updateFields()
{
    if (!nitf_TREUtils_fillData(
                getNative(),
                ((nitf_TREPrivateData*)getNative()->priv)->description,
                &error))
    {
        throw NITFException(&error);
    }
}

nitf::Field TRE::operator[](const std::string& key)
{
    return getField(key);
}

bool TRE::exists(const std::string& key) const
{
    return nitf_TRE_exists(getNativeOrThrow(), key.c_str()) == NITF_SUCCESS;
}

size_t TRE::getCurrentSize() const
{
    const int size = nitf_TRE_getCurrentSize(getNativeOrThrow(), &error);
    if (size < 0)
        throw nitf::NITFException(&error);
    return gsl::narrow<size_t>(size >= 0 ? size : 0);
}

std::string TRE::getTag() const
{
    return static_cast<const char*>(getNativeOrThrow()->tag);
}

void TRE::setTag(const std::string& value)
{
    auto pTag = static_cast<char*>(getNativeOrThrow()->tag);
    memset(pTag, 0, 7);
    memcpy(pTag, value.c_str(), 7);
}

nitf::List TRE::find(const std::string& pattern) const
{
    nitf_List* list = nitf_TRE_find(getNative(), pattern.c_str(), &error);
    if (!list)
        throw except::NoSuchKeyException();
    return nitf::List(list);
}

std::string TRE::getID() const
{
    const char* id = nitf_TRE_getID(getNativeOrThrow());
    return id ? std::string(id) : "";
}

static bool endsWith(const std::string& s, const std::string& match) noexcept
{
    const size_t mLen = match.length();
    const size_t sLen = s.length();
    for (size_t i = 0; i < sLen && i < mLen; ++i)
        if (!(s[sLen - i - 1] == match[mLen - i - 1]))
            return false;
    return sLen >= mLen;
}

std::string TRE::truncate(const std::string& value, size_t maxDigits) const
{
    const size_t decimalIndex = value.find('.');
    if (decimalIndex == std::string::npos)
    {
        return value;
    }

    const std::string truncated = value.substr(0, maxDigits);
    if (endsWith(truncated, "."))
    {
        return truncated.substr(0, truncated.size() - 1);
    }
    if (truncated.size() > decimalIndex)
    {
        return truncated;
    }
    return value;
}

static NITF_BOOL TRE_setField(nitf_TRE* tre,
    const std::string& tag,
    const NITF_DATA* data,
    size_t dataLength, nitf_Error& error) noexcept
{
    return nitf_TRE_setField(tre, tag.c_str(), data, dataLength, &error);
}
static NITF_BOOL TRE_setField(nitf_TRE* tre,
    const std::string& tag,
    const std::string& data,
    nitf_Error& error) noexcept
{
    return TRE_setField(tre, tag, data.c_str(), data.size(), error);
}

void TRE::setField(const std::string& key, const std::string& strValue, NITF_DATA* data, size_t dataLength, bool forceUpdate)
{
    const nitf_Field* const field = nitf_TRE_getField(getNative(), key.c_str());
    if (!field)
    {
        std::ostringstream msg;
        msg << key << " is not a recognized field for this TRE";
        throw except::Exception(Ctxt(msg.str()));
    }
    if (field->type == NITF_BINARY)
    {
        if (!TRE_setField(getNative(),
            key,
            data, dataLength,
            error))
        {
            throw NITFException(&error);
        }
    }
    else
    {
        std::string s = truncate(strValue, field->length);
        if (!TRE_setField(getNative(),
            key,
            s,
            error))
        {
            throw NITFException(&error);
        }
    }

    if (forceUpdate)
    {
        updateFields();
    }

}

nitf_TRE* TRE::create(const std::string& tag, const std::string& id, nitf_Error& error) noexcept
{
    const auto pId = id.empty() ? nullptr : id.c_str();
    return nitf_TRE_construct(tag.c_str(), pId, &error);
}

bool TRE::setField(nitf_TRE* tre, const std::string& tag, const std::string& data, nitf_Error& error) noexcept
{
    return TRE_setField(tre, tag, data, error);
}
